File: src/main/kotlin/com/example/Unstable/EqualityComputed.kt
package com.example.Unstable

import com.example.Computed

class EqualityComputed<T>(getter: () -> T) : Computed<T>(getter) {

    override fun update(): Boolean {
        val prevSub = activeSub
        val prevTrackId = activeTrackId
        setActiveSub(this, nextTrackId())
        startTrack(this)
        val oldValue = cachedValue
        val newValue: T
        try {
            newValue = getter()
        } finally {
            setActiveSub(prevSub, prevTrackId)
            endTrack(this)
        }
        if (!equals(oldValue, newValue)) {
            cachedValue = newValue
            version++
            return true
        }
        return false
    }

    private fun equals(a: Any?, b: Any?): Boolean {
        if (a === b) return true
        if (a == null || b == null) return false
        if (a::class != b::class) return false
        if (a is Array<*> && b is Array<*>) {
            if (a.size != b.size) return false
            for (i in a.indices) {
                if (!equals(a[i], b[i])) return false
            }
            return true
        }
        if (a is Collection<*> && b is Collection<*>) {
            if (a.size != b.size) return false
            val iterA = a.iterator()
            val iterB = b.iterator()
            while (iterA.hasNext() && iterB.hasNext()) {
                if (!equals(iterA.next(), iterB.next())) return false
            }
            return true
        }
        if (a is Map<*, *> && b is Map<*, *>) {
            if (a.size != b.size) return false
            for (key in a.keys) {
                if (!b.containsKey(key) || !equals(a[key], b[key])) return false
            }
            return true
        }
        return a == b
    }
}

File: src/unstable/equalityComputed.ts
import { Computed, ISignal } from '../index.js';

export function equalityComputed<T>(getter: () => T): ISignal<T> {
  return new EqualityComputed(getter);
}

export class EqualityComputed<T = any> extends Computed<T> {
  constructor(getter: () => T) {
    super(oldValue => {
      const newValue = getter();
      if (this.equals(oldValue, newValue)) {
        return oldValue!;
      }
      return newValue;
    });
  }

  equals(a: any, b: any): boolean {
    if (a === b) {
      return true;
    }

    if (a === null || b === null || typeof a !== typeof b) {
      return false;
    }

    if (typeof a === 'object') {
      if (Array.isArray(a) && Array.isArray(b)) {
        if (a.length !== b.length) {
          return false;
        }
        for (let i = 0; i < a.length; i++) {
          if (!this.equals(a[i], b[i])) {
            return false;
          }
        }
        return true;
      }

      if (!Array.isArray(a) && !Array.isArray(b)) {
        for (const key in a) {
          if (a.hasOwnProperty(key)) {
            if (!b.hasOwnProperty(key) || !this.equals(a[key], b[key])) {
              return false;
            }
          }
        }
        for (const key in b) {
          if (b.hasOwnProperty(key) && !a.hasOwnProperty(key)) {
            return false;
          }
        }
        return true;
      }

      return false; // One is array and the other is not
    }

    return false;
  }
}
